Mathlib/Tactic/GRW/Core.lean

```diff
@@ -0,0 +1,179 @@
+mvar of type either `p[x/y]` or `p[y/x]` depending on the value of the `rev` parameter
+
+Parameters:
+* `goal` The mvar for the current goal
+* `expr` the expression to rewrite in
+* `rule` An expression of type `x ~ y` where `~` is some relation
+* `rev` if true, we will rewrite `expr` to `newExpr := expr[y/x]`, otherwise `newExpr := expr[x/y]`
+* `isTarget` if true we are proving `newExpr → expr` otherwise we are proving `expr → newExpr`
+
+Returns:
+* `newExpr` The rewritten version of `expr`
+* `proof` if `isTarget` is true this is a value of type `newExpr → expr`,
+  otherwise it has type `expr → newExpr`
+* `subgoals` list of side goals created by `gcongr`. This does not include the goals
+  successfully filled by `gcongr_discharger`
+-/
+def _root_.Lean.MVarId.grw (goal : MVarId) (expr rule : Expr) (rev isTarget : Bool) :
+    MetaM (Expr × Expr × Array MVarId) := do
+  let oldExpr ← instantiateMVars expr
+  let (ruleArgs, _, resultType) ← forallMetaTelescope (← inferType rule)
+  if (← whnfR resultType).isEq then
+    let ⟨newExpr, eqProof, subgoals⟩ ← goal.rewrite expr rule rev
+    let proof ← if isTarget then
+      mkAppOptM ``cast #[newExpr, expr, ← mkEqSymm eqProof]
+    else
+      mkAppOptM ``cast #[expr, newExpr, eqProof]
+    return ⟨newExpr, proof, subgoals.toArray⟩
+  let rule := mkAppN rule ruleArgs
+  let (template, newExpr) ← getNewExpr rule rev oldExpr
+  let lemmas ← labelled `grw
+  let (lhsExpr, rhsExpr) := if isTarget then (newExpr, oldExpr) else (oldExpr, newExpr)
+
+  -- TODO surely this can be faster
+  for lem in lemmas do
+    let lemResult ← withTraceNode `GRW (fun _ ↦ return m!"trying lemma {lem}") do
+      let lemResult ← try commitIfNoEx do
+        let lemExpr ← mkConstWithFreshMVarLevels lem
+        let lemType ← inferType lemExpr
+        let (metas, binders, _) ← withReducible <| forallMetaTelescopeReducing lemType
+        guard <| ← isDefEq rhsExpr (← inferType (mkAppN lemExpr metas))
+
+        withLocalDeclD (← mkFreshUserName `h) lhsExpr fun value => do
+          if let some firstDefaultArg := binders.findIdx? (· == .default) then
+            withReducible <| metas[firstDefaultArg]!.mvarId!.assignIfDefeq value
+          else
+            throwError "Lemma {lem} did not have a default argument"
+
+          trace[GRW] "Lemma {lem} matches, trying to fill args"
+          pure <| some ⟨← mkLambdaFVars #[value] (mkAppN lemExpr metas), metas⟩
+      catch ex => do
+        trace[GRW] "error in lemma {lem}: {ex.toMessageData}"
+        pure none
+
+      if let some (lemExpr, metas) := lemResult then
+        let metas ← metas.filterM fun x => not <$> x.mvarId!.isAssigned
+        let args := template.getAppArgs
+        -- HACK: we are assuming the side goals of the grw lemma come in the same order
+        -- as they appear in the function, and all the fixed args (e.g. type, instances)
+        -- come before the ones to rewrite
```

Comment:
I'm not completely sure what you mean by "as they appear in the function" here. What are the effects of violating this assumption? Can it be checked when tagging a lemma as `@[grw]`?

Reply:
Yes, this was planned to be checked in the attribute. The [HM-grw-attribute](https://github.com/leanprover-community/mathlib4/commit/ef82c2de29ac2f582687c38f809f199ed3373698) branch implements that.